src/fal/planner/plan.py

```diff
@@ -41,16 +67,74 @@ def from_execution_plan(
 
 
 @dataclass
-class PlannedGraph(OriginGraph):
+class ShuffledGraph(OriginGraph):
     graph: nx.DiGraph
+    after_pattern = re.compile("^script\\..+\\.AFTER\\..+\\.(py|ipynb)")
+    before_pattern = re.compile("^script\\..+\\.BEFORE\\..+\\.(py|ipynb)")
 
     @classmethod
     def from_filtered_graph(
         cls,
         filtered_graph: FilteredGraph,
-        enable_chunking: bool = True,
     ):
         graph = filtered_graph.copy_graph()
+        shuffled_graph = cls(graph)
+        shuffled_graph._shuffle()
+        return shuffled_graph
+
+    def _shuffle(self):
+        def _pattern_matching(pattern: re.Pattern, nodes: Set[str]):
+            return {
+                maybe_script for maybe_script in nodes if pattern.match(maybe_script)
+            }
+
+        old_graph = self.copy_graph()
+        node: str
+        for node in old_graph.nodes:
+
+            all_successors = set(old_graph.successors(node))
+            # TODO: is_after_script() later on!!
+            after_scripts = _pattern_matching(
+                ShuffledGraph.after_pattern, all_successors
+            )
+
+            for succ in all_successors - after_scripts:
+                # Keep the original `node` to `succ` edge and add a new one from `script`
+                self.graph.add_edges_from([(script, succ) for script in after_scripts])
+
+                # And add an edge between all node after scripts to the succ before scripts
+                succ_predecessors = old_graph.predecessors(succ)
+                # TODO: is_before_script() later on!!
+                succ_before_scripts = _pattern_matching(
+                    ShuffledGraph.before_pattern, succ_predecessors
+                )
+                for before_script in succ_before_scripts:
```

Comment:
Maybe we can refactor this method to make it more readable. Otherwise we're in a for loop inside a for loop inside a for loop and it's a little tough to wrap your head around it.

Reply:
done